packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -868,6 +881,118 @@ class ExactAssetImage extends AssetBundleImageProvider {
+/// hardware resources for images that will never be shown on screen.
+///
+/// If the image is already in the cache, e.g. from [precacheImage] or because
+/// another provider has already resolved it, the [getNextAction] callback will
+/// never fire and the cached image is reused.
+@immutable
+class DeferringImageProvider<T> extends ImageProvider<T> {
+
+  /// Creates a new [DeferringImageProvider].
+  ///
+  /// All parameters must not be null.
+  const DeferringImageProvider({
+    @required this.imageProvider,
+    @required this.getNextAction,
+  }) : assert(imageProvider != null),
+       assert(getNextAction != null);
+
+  /// The wrapped [ImageProvider] to delegate loading the image and creating
+  /// a key for it.
+  ///
+  /// The image provider's load method may never be called if resolution gets
+  /// cancelled.
+  final ImageProvider<T> imageProvider;
+
+  /// A callback used by [resolve] to determine whether resolution should be
+  /// cancelled, deferred, or executed.
+  ///
+  /// This callback is never called if the [ImageCache] already has the image
+  /// present, such as from [precacheImage] or another provider successfully
+  /// loading it.
+  final DeferringImageProviderAction Function() getNextAction;
+
+  @override
+  void resolveForStream(ImageConfiguration configuration, ImageStream stream) {
+    void deferredResolve(Object key) {
+      if (PaintingBinding.instance.imageCache.containsKey(key)) {
+        imageProvider.resolveForStream(configuration, stream);
+        return;
+      }
+      switch (getNextAction()) {
+        case DeferringImageProviderAction.defer:
+          // Get ourselves to the end of any frame callbacks at the beginning of
+          // the next frame, which should be after the scroll animation has
+          // updated our velocity value.
+          SchedulerBinding.instance.scheduleFrameCallback((_) {
+            scheduleMicrotask(() => deferredResolve(key));
+          });
+          return;
+        case DeferringImageProviderAction.cancel:
+          return;
+        case DeferringImageProviderAction.resolve:
+          imageProvider.resolveForStream(configuration, stream);
+          return;
+      }
+    }
+    obtainKey(configuration).then((Object key) {
+      deferredResolve(key);
+    });
+  }
+
+  @override
+  ImageStreamCompleter load(T key, DecoderCallback decode) => imageProvider.load(key, decode);
+
+  @override
+  Future<T> obtainKey(ImageConfiguration configuration) => imageProvider.obtainKey(configuration);
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is DeferringImageProvider
+        && other.imageProvider == imageProvider
+        && other.getNextAction == getNextAction;
+  }
+
+  @override
+  int get hashCode => hashValues(imageProvider, getNextAction);
+
+  @override
+  String toString() => '$runtimeType(imageProvider: $imageProvider)';
```

Comment:
please don't use runtimeType.toString()

Reply:
it is incredibly slow

Reply:
This is our standard toString template right?

Reply:
(we do this everywhere)

Reply:
That doesn't make it fast

Reply:
Fair - how about we handle this in a separate PR and upate the style guide?

Reply:
(or file a bug against Dart SDK to make it be fast?)

Reply:
it's fine to use `runtimeType.toString`, just don't use it in release builds.

Reply:
This is done.